File: frontend/src/db/seed.ts
/**
 * Seed the exercise library. Idempotent — skips exercises that already exist
 * (matched by name). Run via `pnpm db:seed`.
 */
import { sql } from "drizzle-orm";

import { getDb } from "./index";
import { exercises } from "./schema";
import seedData from "./seed-exercises.json" with { type: "json" };

interface SeedExercise {
  name: string;
  aliases: string[];
  primary_muscles: string[];
  secondary_muscles: string[];
  equipment: string | null;
  movement_pattern: string | null;
  is_compound: boolean;
  notes?: string | null;
}

async function seed() {
  const db = getDb();

  const existing = await db
    .select({ name: exercises.name })
    .from(exercises);
  const existingNames = new Set(existing.map((r) => r.name));

  const rows = (seedData as SeedExercise[])
    .filter((e) => !existingNames.has(e.name))
    .map((e) => ({
      name: e.name,
      aliases: e.aliases ?? [],
      primaryMuscles: e.primary_muscles ?? [],
      secondaryMuscles: e.secondary_muscles ?? [],
      equipment: e.equipment ?? null,
      movementPattern: e.movement_pattern ?? null,
      isCompound: e.is_compound ?? false,
      notes: e.notes ?? null,
    }));

  if (rows.length === 0) {
    console.log(
      `Seed already applied — ${existingNames.size} exercises in the DB.`
    );
    return;
  }

  await db.insert(exercises).values(rows);

  const total = await db.execute<{ count: string }>(
    sql`select count(*)::text as count from exercises`
  );
  console.log(
    `Seeded ${rows.length} new exercises. Total now: ${total.rows[0]?.count ?? "?"}.`
  );
}

seed()
  .then(() => process.exit(0))
  .catch((err) => {
    console.error("Seed failed:", err);
    process.exit(1);
  });

File: backend/seed.py
"""Run this to seed the exercise library."""
import json
from sqlmodel import Session, select
from database import engine, create_db_and_tables
from models import Exercise
from seed_exercises import get_seed_data


def seed():
    create_db_and_tables()
    with Session(engine) as session:
        existing = session.exec(select(Exercise)).all()
        existing_names = {e.name for e in existing}
        
        added = 0
        for ex_data in get_seed_data():
            if ex_data["name"] in existing_names:
                continue
            exercise = Exercise(
                name=ex_data["name"],
                aliases=json.dumps(ex_data.get("aliases", [])),
                primary_muscles=json.dumps(ex_data.get("primary_muscles", [])),
                secondary_muscles=json.dumps(ex_data.get("secondary_muscles", [])),
                equipment=ex_data.get("equipment"),
                movement_pattern=ex_data.get("movement_pattern"),
                is_compound=ex_data.get("is_compound", False),
                notes=ex_data.get("notes"),
            )
            session.add(exercise)
            added += 1
        
        session.commit()
        print(f"Seeded {added} exercises. Total: {added + len(existing_names)}")


if __name__ == "__main__":
    seed()
